akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1252,9 +1213,31 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
     // make sure any invokeWithFeedback after this fails fast
     // and fail current outstanding invokeWithFeedback promises
     val inProgress = asyncCallbacksInProgress.getAndSet(null)
-    inProgress.foreach { case (acb, promises) ⇒ acb.onStop(promises) }
+    val exception = streamDetatchedException
+    inProgress.foreach(_.tryFailure(exception))
+  }
+
+  private[this] var asyncCleanupCounter = 0L
+  private[stream] def onFeedbackDispatched(p: Promise[Done]): Unit = {
+    // thread-unsafe access, must only be invoked from onAsyncCallback handling
+    asyncCleanupCounter += 1
+
+    if (asyncCleanupCounter % 1000 == 0) {
```

Comment:
what if there are only 999 calls, don't we need to cleanup those?

Reply:
This was just an arbitrary number for testing. The idea would be that 999 promises are still just a constant amount of memory. I didn't check but those promises (if already completed) shouldn't keep references to anywhere else. Could still be quite some amount of memory.

Here are performance numbers for other thresholds:

```
16:
SendQueueBenchmark.queue                                   thrpt   30   576.672 ±   19.172  ops/ms

64:
SendQueueBenchmark.queue                                   thrpt   30  843.796 ±  67.816  ops/ms

128:
SendQueueBenchmark.queue                                   thrpt   30  852.743 ±  85.460  ops/ms

256:
SendQueueBenchmark.queue                                   thrpt   30  1065.788 ±  30.443  ops/ms
```

An alternative (or additional) strategy would be to schedule cleanup regularly so that even those last 999 Promises would be freed.

Reply:
ok, it's just `Done` values, and I assume that registered callbacks to these Promises/Futures have been removed when triggered. Then I think this approach is good. Perhaps 100 is enough?

Reply:
sorry, wrote that comment before I saw yours, 256 seems good then

Reply:
Couldn't the promises be keeping lots of stuff in memory though, through callbacks closing over some state where `invokeWithFeedback` is used?